File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/utils/SurfaceContactUtils.java
package com.github.aadvorak.artilleryonline.battle.utils;

import com.github.aadvorak.artilleryonline.battle.common.Contact;
import com.github.aadvorak.artilleryonline.battle.common.lines.*;
import com.github.aadvorak.artilleryonline.battle.model.RoomModel;
import com.github.aadvorak.artilleryonline.battle.state.SurfaceState;

import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

public class SurfaceContactUtils {

    public static Set<Contact> getContacts(BodyPart bodyPart, RoomModel roomModel, boolean withMaxDepth) {
        var surfaces = roomModel.getState().getSurfaces();
        var contacts = new HashSet<Contact>();
        if (surfaces == null || surfaces.isEmpty()) {
            return contacts;
        }
        var maxDepth = withMaxDepth ? roomModel.getSpecs().getGroundMaxDepth() : 0.0;
        surfaces.forEach(surface -> getContact(bodyPart, surface, maxDepth).ifPresent(contacts::add));
        return contacts;
    }

    private static Optional<Contact> getContact(BodyPart bodyPart, SurfaceState surface, double maxDepth) {
        if (surface.getBoundaries().noOverlap(bodyPart.boundaries())) {
            return Optional.empty();
        }
        var contact = ContactUtils.getBodyPartsContact(bodyPart, surface.getTrapeze());
        if (contact != null) {
            if (maxDepth > 0) {
                return Optional.of(Contact.of(contact.depth() - maxDepth,
                        contact.normal(), contact.position()));
            }
            return Optional.of(contact);
        }
        return Optional.empty();
    }
}

File: frontend/playground/utils/surface-contact-utils.ts
import {Contact} from "~/playground/data/common";
import type {BodyPart} from "~/playground/data/geometry";
import type {RoomModel} from "~/playground/data/model";
import {ContactUtils, IntersectionsPolygonsContactDetector} from "~/playground/utils/contact-utils";
import {Trapeze} from "~/playground/data/geometry";
import type {SurfaceState} from "~/playground/data/state";

export const SurfaceContactUtils = {

  getContacts(bodyPart: BodyPart, roomModel: RoomModel, withMaxDepth: boolean): Set<Contact> {
    const surfaces = roomModel.state.surfaces
    const contacts = new Set<Contact>()
    if (!surfaces || surfaces.length === 0) {
      return contacts
    }
    const maxDepth = withMaxDepth ? roomModel.specs.surfaceMaxDepth : 0.0
    surfaces.forEach(surface => {
      const contact = this.getContact(bodyPart, surface, maxDepth)
      if (contact) {
        contacts.add(contact)
      }
    })
    return contacts
  },

  getContact(bodyPart: BodyPart, surface: SurfaceState, maxDepth: number): Contact | null {
    const trapeze = Trapeze.ofSurface(surface)
    const surfaceBoundaries = trapeze.getBoundaries()
    const bodyBoundaries = bodyPart.getBoundaries()
    if (surfaceBoundaries.noOverlap(bodyBoundaries)) {
      return null
    }
    const contact = ContactUtils.getBodyPartsContactWithDetector(bodyPart, trapeze,
        new IntersectionsPolygonsContactDetector())
    if (contact) {
      if (maxDepth > 0) {
        return Contact.withNormal(
          contact.depth - maxDepth,
          contact.normal,
          contact.position
        )
      }
      return contact
    }
    return null
  }
}
